src/main/java/de/tum/in/www1/artemis/service/scheduled/ProgrammingExerciseScheduleService.java

```diff
@@ -89,25 +95,38 @@ public void scheduleRunningExercisesOnStartup() {
     }
 
     /**
-     * Will cancel a scheduled task if the buildAndTestAfterDueDate is null or has passed already.
+     * Will cancel or reschedule tasks for updated programming exercises
      *
-     * // TODO: the method name and logic is really hard to understand, we should improve this
      * @param exercise ProgrammingExercise
      */
     @Override
-    public void scheduleExerciseIfRequired(ProgrammingExercise exercise) {
-        // TODO: also take exercises with manual assessments into account here and deal better with exams
-        if (!isExamExercise(exercise)
-                && (exercise.getBuildAndTestStudentSubmissionsAfterDueDate() == null || exercise.getBuildAndTestStudentSubmissionsAfterDueDate().isBefore(ZonedDateTime.now()))) {
-            // this only cancels a schedule, but does not schedule one
-            scheduleService.cancelScheduledTaskForLifecycle(exercise, ExerciseLifecycle.DUE);
-            scheduleService.cancelScheduledTaskForLifecycle(exercise, ExerciseLifecycle.BUILD_AND_TEST_AFTER_DUE_DATE);
+    public void updateScheduling(ProgrammingExercise exercise) {
+        if (!needsToBeScheduled(exercise)) {
+            // If a programming exercise got changed so that any scheduling becomes unnecessary, we need to cancel all scheduled tasks
+            cancelAllScheduledTasks(exercise);
             return;
         }
-        // exam exercises are always scheduled, course exercises are only scheduled if buildAndTestStudentSubmissionsAfterDueDate is set and in the future
         scheduleExercise(exercise);
     }
 
+    private static boolean needsToBeScheduled(ProgrammingExercise exercise) {
+        // Exam exercises need to be scheduled
+        if (isExamExercise(exercise))
```

Comment:
Could you please add brackets to this and the following if statement?

Reply:
Yes. Strange that the style tests passed then.